File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Core/Bntx/Texture.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.IO;
using MiniToolbox.Core.Bntx;
using MiniToolbox.Core.Bntx;

namespace MiniToolbox.Core.Bntx;

/// <summary>
/// Represents a texture entry inside a BNTX file (read-only).
/// </summary>
[DebuggerDisplay("Texture {Name}")]
public class Texture : IResData
{
	public ChannelType ChannelRed { get; set; }
	public ChannelType ChannelGreen { get; set; }
	public ChannelType ChannelBlue { get; set; }
	public ChannelType ChannelAlpha { get; set; }
	public uint Width { get; set; }
	public uint Height { get; set; }
	public uint MipCount { get; set; }
	public SurfaceFormat Format { get; set; }
	public string Name { get; set; }
	public string Path { get; set; }
	public uint Depth { get; set; }
	public TileMode TileMode { get; set; }
	public uint Swizzle { get; set; }
	public int Alignment { get; set; }
	public uint Pitch { get; set; }
	public Dim Dim { get; set; }
	public SurfaceDim SurfaceDim { get; set; }
	public long[] MipOffsets { get; set; }
	public List<List<byte[]>> TextureData { get; set; }
	public uint textureLayout { get; set; }
	public uint textureLayout2 { get; set; }
	public AccessFlags AccessFlags { get; set; }
	public uint[] Regs { get; set; }
	public uint ArrayLength { get; set; }
	public byte Flags { get; set; }
	public uint ImageSize { get; set; }
	public uint SampleCount { get; set; }
	public int ReadTextureLayout { get; set; }
	public int sparseBinding { get; set; }
	public int sparseResidency { get; set; }
	public uint BlockHeightLog2 { get; set; }

	void IResData.Load(BntxFileLoader loader)
	{
		loader.CheckSignature("BRTI");
		loader.LoadHeaderBlock();
		Flags = loader.ReadByte();
		Dim = loader.ReadEnum<Dim>(strict: true);
		TileMode = loader.ReadEnum<TileMode>(strict: true);
		Swizzle = loader.ReadUInt16();
		MipCount = loader.ReadUInt16();
		SampleCount = loader.ReadUInt32();
		Format = loader.ReadEnum<SurfaceFormat>(strict: false);
		AccessFlags = loader.ReadEnum<AccessFlags>(strict: false);
		Width = loader.ReadUInt32();
		Height = loader.ReadUInt32();
		Depth = loader.ReadUInt32();
		ArrayLength = loader.ReadUInt32();
		textureLayout = loader.ReadUInt32();
		textureLayout2 = loader.ReadUInt32();
		byte[] array = loader.ReadBytes(20);
		ImageSize = loader.ReadUInt32();
		if (ImageSize == 0)
		{
			throw new Exception("Empty image size!");
		}
		Alignment = loader.ReadInt32();
		uint num = loader.ReadUInt32();
		SurfaceDim = loader.ReadEnum<SurfaceDim>(strict: false);
		Name = loader.LoadString(null, 0L);
		long num2 = loader.ReadInt64();
		long value = loader.ReadInt64();
		long value2 = loader.ReadInt64();
		long num3 = loader.ReadInt64();
		long num4 = loader.ReadInt64();
		long num5 = loader.ReadInt64();
		// Skip UserDataDict and UserData (we don't need them)
		loader.LoadDict();
		MipOffsets = loader.LoadCustom(() => loader.ReadInt64s((int)MipCount), value);
		ChannelRed = (ChannelType)(num & 0xFF);
		ChannelGreen = (ChannelType)((num >> 8) & 0xFF);
		ChannelBlue = (ChannelType)((num >> 16) & 0xFF);
		ChannelAlpha = (ChannelType)((num >> 24) & 0xFF);
		TextureData = new List<List<byte[]>>();
		ReadTextureLayout = Flags & 1;
		sparseBinding = Flags >> 1;
		sparseResidency = Flags >> 2;
		BlockHeightLog2 = textureLayout & 7;
		int num6 = 0;
		for (int num7 = 0; num7 < ArrayLength; num7++)
		{
			List<byte[]> list = new List<byte[]>();
			for (int num8 = 0; num8 < MipCount; num8++)
			{
				int num9 = (int)((MipOffsets[0] + ImageSize - MipOffsets[num8]) / ArrayLength);
				using (loader.TemporarySeek(num6 + MipOffsets[num8], SeekOrigin.Begin))
				{
					list.Add(loader.ReadBytes(num9));
				}
				if (list[num8].Length == 0)
				{
					throw new Exception($"Empty mip size! Texture {Name} ImageSize {ImageSize} mips level {num8} sizee {num9} ArrayLength {ArrayLength}");
				}
			}
			TextureData.Add(list);
			num6 += list[0].Length;
		}
		int num10 = 0;
		long num11 = MipOffsets[0];
		long[] mipOffsets = MipOffsets;
		foreach (long num13 in mipOffsets)
		{
			MipOffsets[num10++] = num13 - num11;
		}
	}

}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Api/src/bntx/Texture.ts
import { BinaryDataReader, SeekTask } from './BinaryDataReader.js';
import { IResData } from './types.js';
import { SurfaceFormat, TileMode, Dim, SurfaceDim, AccessFlags, ChannelType } from './GFX.js';

export class Texture implements IResData {
    public channelRed: ChannelType = ChannelType.Red;
    public channelGreen: ChannelType = ChannelType.Green;
    public channelBlue: ChannelType = ChannelType.Blue;
    public channelAlpha: ChannelType = ChannelType.Alpha;

    public width: number = 0;
    public height: number = 0;
    public mipCount: number = 0;
    public format: SurfaceFormat = SurfaceFormat.Invalid;
    public name: string = "";
    public depth: number = 0;
    public tileMode: TileMode = TileMode.Default;
    public swizzle: number = 0;
    public alignment: number = 0;
    public dim: Dim = Dim.Undefined;
    public surfaceDim: SurfaceDim = SurfaceDim.Dim2D;
    public mipOffsets: number[] = [];
    public textureData: Buffer[][] = []; // [ArrayLayer][MipLevel]
    public arrayLength: number = 1;
    public flags: number = 0;
    public imageSize: number = 0;
    public sampleCount: number = 0;
    public blockHeightLog2: number = 0;

    load(reader: BinaryDataReader): void {
        const sig = reader.readStringWithLength(4);
        if (sig !== 'BRTI') {
            throw new Error(`Invalid Texture signature: ${sig}`);
        }

        // Header block (size/offset)
        const headerOffset = reader.readUInt32();
        const headerSize = reader.readInt64();

        this.flags = reader.readByte();
        this.dim = reader.readByte() as Dim;
        this.tileMode = reader.readUInt16() as TileMode;
        this.swizzle = reader.readUInt16();
        this.mipCount = reader.readUInt16();
        this.sampleCount = reader.readUInt32();
        this.format = reader.readUInt32() as SurfaceFormat;
        const accessFlags = reader.readUInt32() as AccessFlags;

        this.width = reader.readUInt32();
        this.height = reader.readUInt32();
        this.depth = reader.readUInt32();
        this.arrayLength = reader.readUInt32();

        const textureLayout = reader.readUInt32();
        const textureLayout2 = reader.readUInt32();
        reader.seek(20); // Skip reserved

        this.imageSize = reader.readUInt32();
        this.alignment = reader.readInt32();
        const channelMapping = reader.readUInt32();
        this.surfaceDim = reader.readUInt32() as SurfaceDim;

        // Name offset
        const nameOffset = Number(reader.readInt64());

        // Offsets to sub-resources
        const parentBntxOffset = reader.readInt64();
        const mipOffsetsOffset = reader.readInt64();
        const userDataDictOffset = reader.readInt64();
        const userDataOffset = reader.readInt64();
        const textureDataOffset = reader.readInt64();
        const unknownOffset = reader.readInt64();

        // Map channels
        this.channelRed = (channelMapping & 0xFF) as ChannelType;
        this.channelGreen = ((channelMapping >> 8) & 0xFF) as ChannelType;
        this.channelBlue = ((channelMapping >> 16) & 0xFF) as ChannelType;
        this.channelAlpha = ((channelMapping >> 24) & 0xFF) as ChannelType;

        this.blockHeightLog2 = textureLayout & 7;

        // Load Mip Offsets
        if (mipOffsetsOffset !== 0) {
            SeekTask.run(reader, Number(mipOffsetsOffset), 'begin', () => {
                this.mipOffsets = [];
                for (let i = 0; i < this.mipCount; i++) {
                    this.mipOffsets.push(reader.readInt64());
                }
            });
        }

        // Load Name
        if (nameOffset !== 0) {
            SeekTask.run(reader, nameOffset, 'begin', () => {
                // String in BNTX are WordLengthPrefix
                const len = reader.readUInt16();
                this.name = reader.readStringWithLength(len);
            });
        }

        // Load Texture Data
        this.textureData = [];
        if (this.mipOffsets.length > 0) {
            let currentArrayOffset = 0;
            for (let i = 0; i < this.arrayLength; i++) {
                const layerData: Buffer[] = [];
                for (let j = 0; j < this.mipCount; j++) {
                    // C# Logic for calculating mip size:
                    // int num9 = (int)((MipOffsets[0] + ImageSize - MipOffsets[num8]) / ArrayLength);
                    const firstMipOffset = this.mipOffsets[0];
                    const mipSize = Math.floor((firstMipOffset + this.imageSize - this.mipOffsets[j]) / this.arrayLength);

                    SeekTask.run(reader, currentArrayOffset + this.mipOffsets[j], 'begin', () => {
                        layerData.push(reader.readBytes(mipSize));
                    });
                }
                this.textureData.push(layerData);
                // This is a bit simplified, C# updates num6 based on list[0].length
                currentArrayOffset += layerData[0].length;
            }
        }
    }
}
